### PR TITLE
expose addImportMap

### LINKED ISSUE #2428: expose applyImportMap for system.js
### Description of Proposed Feature

system-node.js has a applyImportMap export, for program controlled import make, and api is easier that insert a tag and wait prepareImport 

### In what way would you use it?

like there is already systemJSPrototype.prepareImport, how about systemJSPrototype.applyImportMap=extendImportMap ?

### PR DESCRIPTION
**Pull request by wenerme**
fix #2428

### CODE REVIEW

**Comment by wenerme** [comment]
I don't get the test error, can't find any error messages, running in local seems fine.

**Comment by guybedford** [comment]
The tests seem to be failing here, can you look into it?

**Comment by wenerme** [comment]
@guybedford Sorry, but I have no idea why test failed, it's like firefox failed or something, can you have a look ?

**Comment by guybedford** [comment]
@wenerme for local testing, run `chomp test:browser`. In the Firefox window I'm seeing this error:

image

**Comment by wenerme** [comment]
Running browser test in macos

```bash
CI_BROWSER=/Applications/Firefox.app/Contents/MacOS/firefox chomp test
```

**Review by guybedford — CHANGES_REQUESTED** [review]
This is looking good, we just need to add some documentation on the method as well to the SystemJS API documentation.

Also, in terms of the API name, how about `addImportMap` instead of `applyImportMap`? That would also match the corresponding methods in the ES module shims project (https://github.com/guybedford/es-module-shims#reading-current-import-map-state)

**Comment by wenerme** [comment]
`add` more precisely, renamed to addImportMap, add some doc about addImportMap

File: docs/import-maps.md [hunk 1]
```diff
@@ -200,6 +200,18 @@ Any existing mappings are replaced, although in future this may be an error.
 
 Previous versions of the import maps spec had support for multiple import maps in a single web page ([explanation](https://github.com/WICG/import-maps/issues/199)). SystemJS added support for multiple import maps during that time and has decided to keep support for multiple import maps as an experimental feature ([discussion](https://github.com/systemjs/systemjs/issues/2095)). Note that the Chrome implementation of import maps does not yet allow for multiple maps, and use of multiple import maps within SystemJS should be considered experimental and subject to change.
 
+### Dynamically add Import Maps
+
+Standard import map is based on dom operation, for SystemJS, you can use `System.addImportMap` to add import map dynamically for browser and node.
```
Comment (guybedford) [inline docs/import-maps.md:205]: Perhaps something like:

```suggestion
To dynamically add import maps, a non-spec API is available to extend additional mappings into the import map at any time:
```

**Review by guybedford — APPROVED** [review]
Looks great, thanks for working on this.

Note in es module shims we explicitly avoid overrides when doing this per https://github.com/guybedford/es-module-shims#overriding-import-map-entries, with overrides explicily enabled via configuration. Worth noting but perhaps SystemJS doesn't need this as much.

File: docs/api.md [hunk 2]
```diff
@@ -139,3 +139,16 @@ for (const [id, ns] of System.entries()) {
   console.log(ns); // { exportName: 'value' }
 };
 ```
+
+#### System.addImportMap(map [, base])
+Type: `Function`
+
+Allows you to add import-map without DOM interop.
```
Comment (guybedford) [inline docs/api.md:146]:
```suggestion
Allows adding an import map without using the DOM.
```

File: docs/import-maps.md [hunk 3]
```diff
@@ -200,6 +200,22 @@ Any existing mappings are replaced, although in future this may be an error.
 
 Previous versions of the import maps spec had support for multiple import maps in a single web page ([explanation](https://github.com/WICG/import-maps/issues/199)). SystemJS added support for multiple import maps during that time and has decided to keep support for multiple import maps as an experimental feature ([discussion](https://github.com/systemjs/systemjs/issues/2095)). Note that the Chrome implementation of import maps does not yet allow for multiple maps, and use of multiple import maps within SystemJS should be considered experimental and subject to change.
 
+### Dynamically add Import Maps
+
+> **Note** Non-standard Extension
+> 
+> This may override existing import maps, for more details, please check https://github.com/guybedford/es-module-shims#overriding-import-map-entries
```
Comment (guybedford) [inline docs/import-maps.md:207]:
```suggestion
> Non-standard Extension
```

File: docs/import-maps.md [hunk 4]
```diff
@@ -200,6 +200,22 @@ Any existing mappings are replaced, although in future this may be an error.
...
+
+To dynamically add import maps, a non-spec API is available to extend additional mappings into the import map at any time:
```
Comment (guybedford) [inline docs/import-maps.md:209]:
```suggestion
The `addImportMap` method is available to dynamically extend additional mappings into the import map at any time:
```

File: docs/import-maps.md [hunk 5]
```diff
@@ -200,6 +200,22 @@ Any existing mappings are replaced, although in future this may be an error.
...
+
+```js
+System.addImportMap({
+  "imports": {
+    "y": "/path/to/y.js",
+  }
+})
+```
```
Comment (guybedford) [inline docs/import-maps.md:217]:
```suggestion
```

Any existing map entries will be overridden with the new values.
```

File: docs/import-maps.md [hunk 6]
```diff
@@ -200,6 +200,22 @@ Any existing mappings are replaced, although in future this may be an error.
...
+
```
Comment (guybedford) [inline docs/import-maps.md:218]:
```suggestion

```suggestion

Any existing map entries will be overridden with the new values.
```

**Pull request merged**